### ISSUE TITLE
Testsuite failing with mock errors on Debian with 9.24.24

### ISSUE THREAD

**Issue by glaubitz** [OSInside/kiwi#2072]
I just tried updating kiwi in Debian from 9.2.20 to 9.2.24 which resulted in a lot of testsuite failures related to `mock`:

```
______ ERROR at setup of TestArchiveTar.test_invalid_tar_command_version _______

args = (<unit.archive.tar_test.TestArchiveTar object at 0x7f64bf1f0ee0>,)
keywargs = {}
newargs = (<unit.archive.tar_test.TestArchiveTar object at 0x7f64bf1f0ee0>, <MagicMock name='run' id='140070680007344'>)
newkeywargs = {}

    @wraps(func)
    def patched(*args, **keywargs):
        with self.decoration_helper(patched,
                                    args,
                                    keywargs) as (newargs, newkeywargs):
>           return func(*newargs, **newkeywargs)
E           TypeError: setup() missing 1 required positional argument: 'mock_command'

/usr/lib/python3/dist-packages/mock/mock.py:1346: TypeError
________________ ERROR at setup of TestArchiveTar.test_extract _________________

args = (<unit.archive.tar_test.TestArchiveTar object at 0x7f64bf205d60>,)
keywargs = {}
newargs = (<unit.archive.tar_test.TestArchiveTar object at 0x7f64bf205d60>, <MagicMock name='run' id='140070680026224'>)
newkeywargs = {}

    @wraps(func)
    def patched(*args, **keywargs):
        with self.decoration_helper(patched,
                                    args,
                                    keywargs) as (newargs, newkeywargs):
>           return func(*newargs, **newkeywargs)
E           TypeError: setup() missing 1 required positional argument: 'mock_command'

/usr/lib/python3/dist-packages/mock/mock.py:1346: TypeError
_________________ ERROR at setup of TestArchiveTar.test_create _________________

args = (<unit.archive.tar_test.TestArchiveTar object at 0x7f64bf00dcd0>,)
keywargs = {}
newargs = (<unit.archive.tar_test.TestArchiveTar object at 0x7f64bf00dcd0>, <MagicMock name='run' id='140070677962080'>)
newkeywargs = {}

    @wraps(func)
    def patched(*args, **keywargs):
        with self.decoration_helper(patched,
                                    args,
                                    keywargs) as (newargs, newkeywargs):
>           return func(*newargs, **newkeywargs)
E           TypeError: setup() missing 1 required positional argument: 'mock_command'

/usr/lib/python3/dist-packages/mock/mock.py:1346: TypeError
```

Full build log available in: https://people.debian.org/~glaubitz/kiwi_9.24.24-1_amd64.build

I haven't tried debugging this issue yet, but maybe anyone else has a clue what might the problem, i.e. maybe due to an additional build dependency that's now required.

**Comment by schaefi** [comment]
This is related to a change done on the pytest side. See #2058

**Comment by schaefi** [comment]
I wonder why it still happens to you.... will take a look

**Comment by schaefi** [comment]
ok I think I found it. Houston we have a problem

If you look at the test run in the github actions you see:

* https://github.com/OSInside/kiwi/runs/5264813580?check_suite_focus=true

and in the log you see:

```
pytest==7.0.1
```

With the move to 7.0.1 there was an api change in pytest which required us to update the test code
as done in #2058 

The issue is now that the tests fail on older versions of pytest. I have kiwi for debian building in obs
here:

https://build.opensuse.org/package/live_build_log/home:marcus.schaefer/kiwi-debian/Debian_Unstable/x86_64

it uses:

```
[   55s] [266/267] installing python3-pytest-6.2.5-1
```

from Debian unstable, and that's why it fails

**Comment by schaefi** [comment]
I think we have two options

1. We stick with pytest 6.x in tox.ini and revert back all changes we did to support pytest 7.x
2. We make pytest 7.x available in Debian, but also 6.x has to be available there due to the api changes to prevent other peoples test to break

Other options ideas ?

@davidcassany @Conan-Kudo I need your input with that part

Thanks

**Comment by Conan-Kudo** [comment]
We'll need to support both 6.x and 7.x in parallel no matter what, as eventually Fedora Rawhide will move forward to 7.x (I expect @hroncok will upgrade Fedora to pytest 7.x soon in Rawhide), so we'll get bitten _again_ if we don't adapt for 7.x while supporting 6.x.

**Comment by davidcassany** [comment]
Ouch this is an ugly issue... It looks really unfortunate we can't define a 6.x and 7.x compatible tests. I believe we should double check if we can figure out a way to make tests compatible for both versions (I guess you might have had a look at it already, I am new to this issue) and if not I'd say better stick to v6 for the time being as v7 is really recent.

There is another option I dislike, but still an option, which is not running tests for Debian packages at build time. Given that pytest is not a runtime dependency it should be possible to build Debian packages without pytest at all.

Just my two cents.

**Comment by glaubitz** [comment]
> On Feb 21, 2022, at 1:47 PM, David Cassany Viladomat ***@***.***> wrote:
> 
> There is another option I dislike, but still an option, which is not running tests for Debian packages at build time. Given that pytest is not a runtime dependency it should be possible to build Debian packages without pytest at all.
> 

If you send me a patch to switch the testsuite back to pytest 6.x, I could carry the patch until Debian unstable has switched to pytest 7.x.

Adrian

**Comment by schaefi** [comment]
> We'll need to support both 6.x and 7.x in parallel no matter what, as eventually Fedora Rawhide will move forward to 7.x (I expect @hroncok will upgrade Fedora to pytest 7.x soon in Rawhide), so we'll get bitten _again_ if we don't adapt for 7.x while supporting 6.x.

Supporting both at the same time is imho not easily possible. The setup() routine has changed its parameter list and we would basically need to write a decorator that calls setup() differently depending on the used pytest version. I think it's not worth the effort to do this for the following reasons:

* the current concept of unit testing is based on using a unit test framework from a CI system (github actions) to run the tests for the supported python version (3.6, 3.8, 3.9 and so on) including all the setup.py required python modules. If the tests passes we can make the statement: works on this python with the modules pulled in from pypi and as requested through setup.py. We cannot make the statement: the implementation works.

* For testing the actual implementation on the desired distributions, Fedora, Debian, SUSE, etc etc we maintain this build test matrix on OBS and also run the boot tests through openQA. At this point we also detect inconsistencies caused by python module packages used on the distribution compared to modules used when the unit tests run. If everything works we don't care about different versions of python modules as we also only have limited influence on the packages as they are provided by the distributions. In case we detect an inconsistency we have to move and find a solution.

This is also basically the reason why we don't run the unit tests at build time of the package. From my perspective all required testing was done before the release of a new package and we never had issues with this process for which a test run at package build time would have helped us.

The Debian package as it gets released to Debian unstable however, was done with unit tests to run at package build time.
iirc @glaubitz wanted it that way and it probably also makes sense for the unstable target for which we don't have an integration test run (maybe we should think of adding it as we also test tumbleweed and rawhide rolling releases)

This is why the pytest inconsistency only hurts us on the debian package currently. This is also why there will be no issue if other distributions move to pytest v7 as it's not a runtime dependency for kiwi to continue to work or the package to continue to build.

> There is another option I dislike, but still an option, which is not running tests for Debian packages at build time.

I assume this will not be accepted by the Debian people because @glaubitz spent some time to package this all up and get the approval. Also with unstable not part of the integration test matrix it is probably better to keep the current package design

> and if not I'd say better stick to v6 for the time being as v7 is really recent

I'm leaning towards this idea as it will work for all and is only causing an issue if pytest v6 gets dropped which I don't think is going to happen as it will kill many project test setups.

I created a pull request in that direction and a build test for the Debian package here:

* https://build.opensuse.org/package/show/home:marcus.schaefer/kiwi-debian

Thoughts ?
